core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -0,0 +1,413 @@
+          case _: QueryResult | _: BufferedEvents =>
+            tryPushOne()
+          case _: QueryInProgress =>
+        }
+      }
+
+      private def query(): Unit = {
+        rowsForCurrentQuery = 0
+        log.debug("Executing query: timeBucket: {} offset: {}", currTimeBucket, currentOffset)
+        queryState = QueryInProgress(System.nanoTime())
+        session.selectEventsForBucket(currTimeBucket, currentOffset, toOffset).onComplete(newResultSetCb.invoke)
+      }
+
+      private def lookForMissing(): Unit = {
+        val missing = missingLookup match {
+          case Some(m) => m
+          case None    => throw new IllegalStateException(s"lookingForMissingCalled when there is no missing. Raise a bug with debug logging.")
+        }
+
+        if (missing.deadline.isOverdue()) {
+          log.info("{}: Failed to find missing sequence nr: {}", session.tag, missingLookup)
+          if (missing.failIfNotFound) {
+            fail(out, new IllegalStateException(s"Unable to find missing tagged event: PersistenceId: ${missing.persistenceId}. " +
+              s"TagPidSequenceNr: ${missing.missing}. Previous offset: ${missing.previousOffset}"))
+          } else {
+            stopLookingForMissing(missing, missing.buffered)
+            tryPushOne()
+          }
+        } else {
+          queryState = QueryInProgress()
+          log.info("{}: Executing query to look for missing. Timebucket: {}. From: {}. To: {}", session.tag, missing.bucket, missing.previousOffset, missing.maxOffset)
+          session.selectEventsForBucket(missing.bucket, missing.previousOffset, Some(missing.maxOffset))
+            .onComplete(newResultSetCb.invoke)
+        }
+      }
+
+      @tailrec def tryPushOne(): Unit = {
+        log.debug("TryPushOne() {} {}", queryState, missingLookup)
+        queryState match {
+          case QueryResult(rs) if isAvailable(out) =>
+            if (rs.isExhausted) {
+              queryExhausted()
+            } else if (rs.getAvailableWithoutFetching == 0) {
+              log.debug("Fetching more")
+              fetchMore(rs)
+            } else if (missingLookup.isDefined) {
+              val m = missingLookup.get
+              val event = extractEvent(rs.one())
+              if (event.persistentRepr.persistenceId == m.persistenceId && m.missing.contains(event.tagPidSequenceNr)) {
+                val remainingEvents = m.missing - event.tagPidSequenceNr
+                log.info("{}: Found a missing event, sequence nr {}. Remaining missing: {}", session.tag, event.tagPidSequenceNr, remainingEvents)
+                if (remainingEvents.isEmpty) {
+                  stopLookingForMissing(m, event :: m.buffered)
+                } else {
+                  log.info("{}: There are more missing events. {}", session.tag, remainingEvents)
+                  missingLookup = missingLookup.map(m => m.copy(
+                    missing = remainingEvents,
+                    buffered = event :: m.buffered
```

Comment:
Not explicitly. It will never be bigger than a sequence nr gap that we've detected which is expected to be 1 or a handful most of the time.

We could fail fast if a massive gap is detected? Or we could flip between searching / deliverying chunks of missed events but I feel that is unlikely and it'll add some complexity to the stage

Reply:
Would be nice to fail the stream if too many, to avoid risk of OME. Can be separate ticket.

Reply:
https://github.com/akka/akka-persistence-cassandra/issues/266